### PR TITLE
Add API to list resources under namespace tree

### LINKED ISSUE #235: Support for "hns get pods -n my-hierarchical-ns"?
Hi there. I was excited to find hns today as a potential solution for some multi-tenancy queries users have which cannot be easily solved with vanilla k8s. There are two related requests that pop up in different scenarios for our application ([Kubeapps](https://kubeapps.dev/)):
- Can I see just the namespaces to which I have access? We currently solve this with a special service account that can list namespaces and then check to which namespaces a user has access (not great and expensive).
- Can I see all resource-x's in the namespaces to which I have access... effectively `kubectl hns get pods -n my-hierarchical-ns`

I can see ways that I could use hns to potentially answer the first point above, though I also note that the result [may be slow for the same reason](#223). It seems that currently this is answered client-side (by querying for the anchors in a namespace, then for those namespaces, recursively?).

So my questions are:
1. Are there plans, or has there been other requests, to move the hierarchical query server-side so that a single query can return the available namespaces in the tree? (I think you answered this with a "no" only 22 days ago on #223 :) ), and
2. If (1) was a possibility, could that be easily generalised to answer queries for resources within a hierarchy (eg. `kubectl hns get pods -n myhierarchical-ns`) on the server-side, rather than recursively visiting all namespaces and calling `get pods` etc. (assuming a user without read permission on the entire cluster).

I'm not certain, but it looks like doing so could not only solve the slow responses for `hns tree` but also allow users in a multi-tenant environment to see all their resources (similar to `kubectl get pods -A` but for a hierarchy rather than the whole cluster), which I imagine would be very useful to others as well?

Thanks for any thoughts (and for the great work on hns - I was very excited to find it... been a missing piece in k8s for a long time that Rancher and openshift have tried to fill in the past, imo).

### PR DESCRIPTION
**Pull request by cmurphy**
This change adds a new command and configuration to register an API endpoint which will allow users to query for resources belonging to descendants of any namespace. This means that finding all resources under a parent namespace can be done in one request, instead finding and iterating over all descendants. The new endpoints are:

* /apis/resources.hnc.x-k8s.io/v1alpha1
 - discovery endpoint
* /apis/resources.hnc.x-k8s.io/v1alpha1/{resource}
 - global resource query, equivalent to the same global request for the original resource
* /apis/resources.hnc.x-k8s.io/v1alpha1/namespaces/{namespace}/{resource}
 - returns resources in all namespaces under {namespace}, including {namespace}.

Design: https://docs.google.com/document/d/1WpnAJ3442v93G4Wi7SnoEiLZVLJJO-AWCkl5GO4YLl4/edit?usp=sharing
Fixes: https://github.com/kubernetes-sigs/hierarchical-namespaces/issues/235

This is based on the POC in https://github.com/cmurphy/hns-list.

To do:
- [x] Add the CLI plugin
- [x] make it work with not-cert-manager (default-cc.yaml)
- [x] conform to kustomize patterns
- [x] add support for Table formatted output https://kubernetes.io/docs/reference/using-api/api-concepts/#receiving-resources-as-tables
- [x] unit tests
- [x] integration tests
- [ ] docs

### CODE REVIEW

**Review by adrianludwin — COMMENTED** [review]
Just some early comments. I understand and like the overall approach, thanks! I'm not sure what code is boilerplate and what's not, can you maybe add some comments to indicate?

The only thing I'm not seeing (at a quick review) are the RBAC checks - how do we know that the user is authorized to list resource X in namespace Y? (It is safe to assume that this also gives you access to all descendants of Y).

Thanks for this, looks good so far!

**Review by cmurphy — COMMENTED** [review]
> I'm not sure what code is boilerplate and what's not, can you maybe add some comments to indicate?

I've split the vendor code and the kustomize yaml into separate commits for easier reviewing. Let me know if you'd ultimately like to keep them split or squash them back into one once this is ready to merge.

> The only thing I'm not seeing (at a quick review) are the RBAC checks - how do we know that the user is authorized to list resource X in namespace Y? (It is safe to assume that this also gives you access to all descendants of Y).

Good catch, this is intentional for the moment and an open question in [the design doc](https://docs.google.com/document/d/1WpnAJ3442v93G4Wi7SnoEiLZVLJJO-AWCkl5GO4YLl4/edit#heading=h.9p2at89crnbl). As it is right now, k8s blocks users from using the new resource endpoints unless they are explicitly granted permission with a role like `kubectl -n parentns create role hnc-pods --verb list,watch --resource pods.resources.hnc.x-k8s.io`. This is the least opinionated path and allows for the finest granularity of permission delegation. But it's obviously also a little user-hostile.

There's nothing that can be done from inside the handlers about this because kube-api won't even let the user touch the endpoint without this permission. So one option would to add a routine that auto-creates a catch-all role and rolebinding in each namespace that allows `system:authenticated` users to access all these `*.resources.hnc.x-k8s.io` endpoints, then have the handlers perform the real RBAC check for the target resource. This would make the new APIs "just work" but it's a little extra complexity and it moves the security boundary away from kube-api and into HNC, which might not be every user's preference.

**Comment by adrianludwin** [comment]
LMK when you want me to review this again, thanks for your work on it!

**Comment by cmurphy** [comment]
@adrianludwin thanks for your patience! It's been a busy summer. I'm working on unit tests and then this should be ready for review.

**Comment by adrianludwin** [comment]
Lol I'm replying to some of these issues months late, no need to explain or
apologize!

**Comment by cmurphy** [comment]
Marking this ready for review. It still needs integration tests, but enough of the core functionality is there to get feedback on. Happy to make any changes needed.

The commits are divided by components:

1. the main server functionality
2. go.mod and vendor updates
3. client
4. kustomize
5. cert setup

I know it's a massive code dump, let me know if there's any way I can organize it to make it easier to review. Happy to jump on a call to work through it as well.

Some open questions:
1. This is currently deployed in hnc-system, should it be in a separate namespace?
2. How and whether to handle RBAC - see the [RBAC](https://docs.google.com/document/d/1WpnAJ3442v93G4Wi7SnoEiLZVLJJO-AWCkl5GO4YLl4/edit#heading=h.9p2at89crnbl) section of the design doc
3. Question on the API version https://github.com/kubernetes-sigs/hierarchical-namespaces/pull/281#discussion_r1227222714

**Comment by cmurphy** [comment]
Bump - this is ready for review and includes integration tests now. Let me know if you'd like to get on a call to walk through it.

**Comment by adrianludwin** [comment]
Sorry, super busy but I'll try to do it this week. Thanks Colleen!

File: cmd/apiextension/main.go [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+	if err != nil {
+		setupLog.Error(err, "could not start watcher")
+		os.Exit(1)
+	}
+	dynamicClient, err := dynamic.NewForConfig(cfg)
+	if err != nil {
+		setupLog.Error(err, "could not get dynammic client for config")
+		os.Exit(1)
+	}
+	dynamicFactory := getDynamicInformerFactory(dynamicClient)
```
[the remark this answers is not in the document]
Comment (adrianludwin) [reply cmd/apiextension/main.go]: lol, yes, K8s was originally written by a bunch of Java developers told to use Go, at least as I've heard it!

**Comment by adrianludwin** [comment]
> > The only thing I'm not seeing (at a quick review) are the RBAC checks - how do we know that the user is authorized to list resource X in namespace Y? (It is safe to assume that this also gives you access to all descendants of Y).
> 
> ... As it is right now, k8s blocks users from using the new resource endpoints unless they are explicitly granted permission with a role like `kubectl -n parentns create role hnc-pods --verb list,watch --resource pods.resources.hnc.x-k8s.io`. This is the least opinionated path and allows for the finest granularity of permission delegation. But it's obviously also a little user-hostile.

I'm fine like this, HNC should prioritize security if you have to make a choice, and as you've shown the admin can always create a Role (or ClusterRole) with this stuff built-in and then assign it as needed.

Just one question, I thought this endpoint worked with _any_ HNC-controlled resource, e.g. you'd give permission to `resources.hnc.x-k8s.io`, not (say) `pods.resources.hnc.x-k8s.io`. Did I get that wrong?

> There's nothing that can be done from inside the handlers about this because kube-api won't even let the user touch the endpoint without this permission. So one option would to add a routine that auto-creates a catch-all role and rolebinding in each namespace that allows `system:authenticated` users to access all these `*.resources.hnc.x-k8s.io` endpoints, then have the handlers perform the real RBAC check for the target resource. This would make the new APIs "just work" but it's a little extra complexity and it moves the security boundary away from kube-api and into HNC, which might not be every user's preference.

Yes, I agree, let's not do this.

**Comment by adrianludwin** [comment]
One high-level comment, and I'm sorry I didn't notice this before - I'm worried this is a bit limiting:

`/apis/resources.hnc.x-k8s.io/v1alpha2/namespaces/{namespace}/{resource}`

It doesn't give us a way to later query by anything other than tree, since "namespace" is such a generic name. How hard would it be at this point to change it to:

`/apis/resources.hnc.x-k8s.io/v1alpha2/tree/{namespace}/{resource}`

Then if we want to have other query types, it's pretty easy to insert them.

**Comment by adrianludwin** [comment]
I've added a lot of nitpicky comments but overall I really like this. I haven't run it myself so it would be really nice if you could generate some output and paste it into this PR so I can see what it looks like in practice (both the raw responses from the endpoint, and the plugin output) since that's an important thing to be able to comment on as well. One other general comment I had was to clearly identify what was basically boilerplate (with sources/inspirations if possible), which makes it much easier to focus on the business logic and makes future archeology easier.

I haven't closely examined the unit tests because I don't think I'm really qualified to understand all the intricacies (and you seem to know what you're doing). Once I know what the plugin output is _supposed_ to look like I'll have a closer look at the integ test.

As for squashing, I usually prefer having a single commit per PR, or at least having each commit be independently buildable (which it's not if you update the dependencies in a separate commit). At most, I might submit this as two commits: one for the server-side stuff (implementation, cert-rotator, config, dependencies) and another for the client-side (kubectl, integ test, any _additional_ dependencies). I'm fine either way though.

I'll admit that I appreciated going through the commits one-by-one because it looked overwhelming at first, but now that I understand what you were doing, I could re-review much faster in the future.

Thanks for this, looking good!

**Comment by cmurphy** [comment]
> Just one question, I thought this endpoint worked with any HNC-controlled resource, e.g. you'd give permission to resources.hnc.x-k8s.io, not (say) pods.resources.hnc.x-k8s.io. Did I get that wrong?

You have this right, you could create a role with rules like

```
rules:
- apiGroups:
  - resources.hnc.x-k8s.io
  resources:
  - '*'
  verbs:
  - list
  - watch
```

or be more fine-grained as in my example

```
rules:
- apiGroups:
  - resources.hnc.x-k8s.io
  resources:
  - secrets
  - otherstuff
  verbs:
  - list
  - watch
```
depending on how selective the cluster admin wants to be.

> How hard would it be at this point to change it to:
> `/apis/resources.hnc.x-k8s.io/v1alpha2/tree/{namespace}/{resource}`

It's technically possible to change the endpoint path but there are a few reasons I don't think it would be a good idea:

1. Kubernetes is pretty strict about what it wants an API endpoint to look like, and it dictates that resources are either namespaced or global, there isn't any third option: https://kubernetes.io/docs/reference/using-api/api-concepts/#resource-uris
2. As a follow-on to (1), clients also expect API paths to be in this format. This makes https://github.com/cmurphy/hierarchical-namespaces/blob/c3c3e5ce76f34c9acb105eec72578377f2bcfe3b/internal/kubectl/get.go#L75 work without customizing a non-standard HTTP client with a custom API path. External clients not part of HNC work the same way.
3. Standard k8s RBAC resources also expect namespaced paths. If I create a Role in namespace `parent1`, that means that clients can access resources in path `/apis/<groupversion>/namespaces/<namespace>/<resource>`. There is no way to create a Role to allow access to an endpoint with `/tree/`.
4. This really is about Kubernetes namespaces, you just have to think of it as "resources in descendants of `<namespace>`". Given Kubernetes' strictness about namespaced and non-namespaced things, I can't think of other query types that would fit here, but if we think of them we could put them in a new API group.

> I've added a lot of nitpicky comments

I'm not seeing any inline comments, is it possible you didn't submit the review?

> it would be really nice if you could generate some output and paste it into this PR so I can see what it looks like in practice

Will add in next comment

> One other general comment I had was to clearly identify what was basically boilerplate (with sources/inspirations if possible)

I've updated it to add some comments on where I've pulled inspiration from in places where I didn't come up with it myself. I no longer have all my research for this in my browser history but I did my best to find sources again. The only parts that I think of as "boilerplate" are things like parsing flags where I've copied from other parts of HNC, and I didn't explicitly comment those parts. Let me know if you want more background on any other parts.

> As for squashing, I usually prefer having a single commit per PR, or at least having each commit be independently buildable (which it's not if you update the dependencies in a separate commit)

I'll squash this into one commit once we think it's close to ready to merge.

**Comment by cmurphy** [comment]
# Output Examples

## Discovery

### API

```
$ kubectl get --raw /apis/resources.hnc.x-k8s.io/v1alpha2 | jq .
{
  "apiVersion": "v1",
  "groupVersion": "resources.hnc.x-k8s.io/v1alpha2",
  "kind": "APIResourceList",
  "resources": [
    {
      "name": "events",
      "singularName": "",
      "namespaced": true,
      "version": "v1",
      "kind": "Event",
      "verbs": [
        "list",
        "watch"
      ],
      "shortNames": [
        "ev"
      ],
      "storageVersionHash": "r2yiGXH7wu8="
    },
    {
      "name": "serviceaccounts",
      "singularName": "",
      "namespaced": true,
      "version": "v1",
      "kind": "ServiceAccount",
      "verbs": [
        "list",
        "watch"
      ],
      "shortNames": [
        "sa"
      ],
      "storageVersionHash": "pbx9ZvyFpBE="
    },
    {
      "name": "secrets",
      "singularName": "",
      "namespaced": true,
      "version": "v1",
      "kind": "Secret",
      "verbs": [
        "list",
        "watch"
      ],
      "storageVersionHash": "S6u1pOWzb84="
    },
    ...
```

### Kubectl

```
$ kubectl api-resources | grep resources.hnc
apps.controllerrevisions                                      resources.hnc.x-k8s.io/v1alpha2        true         ControllerRevision
apps.daemonsets                                  ds           resources.hnc.x-k8s.io/v1alpha2        true         DaemonSet
apps.deployments                                 deploy       resources.hnc.x-k8s.io/v1alpha2        true         Deployment
apps.replicasets                                 rs           resources.hnc.x-k8s.io/v1alpha2        true         ReplicaSet
apps.statefulsets                                sts          resources.hnc.x-k8s.io/v1alpha2        true         StatefulSet
authorization.k8s.io.localsubjectaccessreviews                resources.hnc.x-k8s.io/v1alpha2        true         LocalSubjectAccessReview
autoscaling.horizontalpodautoscalers             hpa          resources.hnc.x-k8s.io/v1alpha2        true         HorizontalPodAutoscaler
batch.cronjobs                                   cj           resources.hnc.x-k8s.io/v1alpha2        true         CronJob
batch.jobs                                                    resources.hnc.x-k8s.io/v1alpha2        true         Job
bindings                                                      resources.hnc.x-k8s.io/v1alpha2        true         Binding
configmaps                                       cm           resources.hnc.x-k8s.io/v1alpha2        true         ConfigMap
coordination.k8s.io.leases                                    resources.hnc.x-k8s.io/v1alpha2        true         Lease
discovery.k8s.io.endpointslices                               resources.hnc.x-k8s.io/v1alpha2        true         EndpointSlice
endpoints                                        ep           resources.hnc.x-k8s.io/v1alpha2        true         Endpoints
events                                           ev           resources.hnc.x-k8s.io/v1alpha2        true         Event
events.k8s.io.events                             ev           resources.hnc.x-k8s.io/v1alpha2        true         Event
hnc.x-k8s.io.hierarchicalresourcequotas          hrq          resources.hnc.x-k8s.io/v1alpha2        true         HierarchicalResourceQuota
hnc.x-k8s.io.hierarchyconfigurations                          resources.hnc.x-k8s.io/v1alpha2        true         HierarchyConfiguration
hnc.x-k8s.io.subnamespaceanchors                 subns        resources.hnc.x-k8s.io/v1alpha2        true         SubnamespaceAnchor
limitranges                                      limits       resources.hnc.x-k8s.io/v1alpha2        true         LimitRange
networking.k8s.io.ingresses                      ing          resources.hnc.x-k8s.io/v1alpha2        true         Ingress
networking.k8s.io.networkpolicies                netpol       resources.hnc.x-k8s.io/v1alpha2        true         NetworkPolicy
persistentvolumeclaims                           pvc          resources.hnc.x-k8s.io/v1alpha2        true         PersistentVolumeClaim
pods                                             po           resources.hnc.x-k8s.io/v1alpha2        true         Pod
podtemplates                                                  resources.hnc.x-k8s.io/v1alpha2        true         PodTemplate
policy.poddisruptionbudgets                      pdb          resources.hnc.x-k8s.io/v1alpha2        true         PodDisruptionBudget
rbac.authorization.k8s.io.rolebindings                        resources.hnc.x-k8s.io/v1alpha2        true         RoleBinding
rbac.authorization.k8s.io.roles                               resources.hnc.x-k8s.io/v1alpha2        true         Role
replicationcontrollers                           rc           resources.hnc.x-k8s.io/v1alpha2        true         ReplicationController
resourcequotas                                   quota        resources.hnc.x-k8s.io/v1alpha2        true         ResourceQuota
secrets                                                       resources.hnc.x-k8s.io/v1alpha2        true         Secret
serviceaccounts                                  sa           resources.hnc.x-k8s.io/v1alpha2        true         ServiceAccount
services                                         svc          resources.hnc.x-k8s.io/v1alpha2        true         Service
storage.k8s.io.csistoragecapacities                           resources.hnc.x-k8s.io/v1alpha2        true         CSIStorageCapacity
```

## Global Resource List

### API

#### List

```
$ kubectl get --raw /apis/resources.hnc.x-k8s.io/v1alpha2/secrets | jq '.items[] | "\(.metadata.namespace)/\(.metadata.name)"'
"child1/child1-s1"
"grandchild1/grandchild1-s1"
"hnc-system/hnc-resourcelist"
"hnc-system/hnc-webhook-server-cert"
"parent1/parent1-s1"
```

#### Watch

(Secret child1-s2 is created in another session after watch is started)

```
$ kubectl get --raw '/apis/resources.hnc.x-k8s.io/v1alpha2/secrets?watch=1&resourceVersion=129224'
{"type":"ADDED","object":{"apiVersion":"v1","kind":"Secret","metadata":{"creationTimestamp":"2023-11-10T01:54:46Z","managedFields":[{"apiVersion":"v1","fieldsType":"FieldsV1","fieldsV1":{"f:type":{}},"manager":"kubectl-create","operation":"Update","time":"2023-11-10T01:54:46Z"}],"name":"child1-s2","namespace":"child1","resourceVersion":"129265","uid":"73b8b2b6-a6f5-4995-aaa8-8b3fad771974"},"type":"Opaque"}}
^C
```

### Kubectl

#### List

```
$ kubectl hns get secrets --all-namespaces
NAMESPACE     NAME                      AGE
child1        child1-s1                 2m6s
child1        child1-s2                 25s
grandchild1   grandchild1-s1            112s
hnc-system    hnc-resourcelist          25h
hnc-system    hnc-webhook-server-cert   25h
parent1       parent1-s1                2m14s
```

#### Watch

(Secret child1-s3 is created in another session after watch is started)

```
$ kubectl hns get secrets --all-namespaces --watch
NAMESPACE     NAME                      AGE
child1        child1-s1                 2m41s
child1        child1-s2                 60s
grandchild1   grandchild1-s1            2m27s
hnc-system    hnc-resourcelist          25h
hnc-system    hnc-webhook-server-cert   25h
parent1       parent1-s1                2m49s
child1   child1-s3   0s
^C
```


## Namespaced (Tree) Resource List

### API

#### List

```
$ kubectl get --raw /apis/resources.hnc.x-k8s.io/v1alpha2/namespaces/parent1/secrets | jq '.items[] | "\(.metadata.namespace)/\(.metadata.name)"'
"child1/child1-s1"
"child1/child1-s2"
"child1/child1-s3"
"grandchild1/grandchild1-s1"
"parent1/parent1-s1
```

#### Watch

(Secret grandchild1-s2 is created in another session after watch is started)

```
$ kubectl get --raw '/apis/resources.hnc.x-k8s.io/v1alpha2/namespaces/parent1/secrets?watch=1&resourceVersion=129444'
{"type":"ADDED","object":{"apiVersion":"v1","kind":"Secret","metadata":{"creationTimestamp":"2023-11-10T01:57:34Z","managedFields":[{"apiVersion":"v1","fieldsType":"FieldsV1","fieldsV1":{"f:type":{}},"manager":"kubectl-create","operation":"Update","time":"2023-11-10T01:57:34Z"}],"name":"grandchild1-s2","namespace":"grandchild1","resourceVersion":"129482","uid":"4956b13d-c838-4575-8318-4a96e3dedfd6"},"type":"Opaque"}}
^C
```

### Kubectl

#### List

```
$ kubectl hns get secrets --namespace parent1
NAMESPACE     NAME             AGE
child1        child1-s1        4m49s
child1        child1-s2        3m8s
child1        child1-s3        2m4s
grandchild1   grandchild1-s1   4m35s
grandchild1   grandchild1-s2   20s
parent1       parent1-s1       4m57s
```

#### Watch

(Secret grandchild1-s3 is created in another session after watch is started)

```
$ kubectl hns get secrets --namespace parent1 --watch
NAMESPACE     NAME             AGE
child1        child1-s1        5m39s
child1        child1-s2        3m58s
child1        child1-s3        2m54s
grandchild1   grandchild1-s1   5m25s
grandchild1   grandchild1-s2   70s
parent1       parent1-s1       5m47s
grandchild1   grandchild1-s3   0s
^C
```

**Review by adrianludwin — COMMENTED** [review]
Uuuuuuuggggh, sorry, forgot to hit submit! I'll reply to your latest comments soon

File: internal/apiextension/handlers/handlers.go [hunk 2]
```diff
@@ -0,0 +1,513 @@
...
+	if len(namespaces) == 0 {
+		watcher, err := client.Watch(ctx, opts)
+		if err != nil {
+			return nil, err
+		}
+		return []watch.Interface{watcher}, nil
+	}
+	watcherChan := make(chan watch.Interface)
+	done := make(chan bool)
+	watchers := make([]watch.Interface, 0)
```
[the remark this answers is not in the document]
Comment (cmurphy) [reply internal/apiextension/handlers/handlers.go:216]: No, Watch() isn't particularly slow. The async is unnecessary, I've removed it.

File: internal/apiextension/handlers/handlers.go [hunk 3]
```diff
@@ -0,0 +1,513 @@
...
+		for r := range resourceVersions {
+			if r > latestResourceVersion {
+				latestResourceVersion = r
+			}
+		}
+		wg.Done()
+	}()
+
+	eg, ctx := errgroup.WithContext(r.Context())
+	sem := semaphore.NewWeighted(workers)
```
[the remark this answers is not in the document]
Comment (cmurphy) [reply internal/apiextension/handlers/handlers.go:327]: Correct

File: internal/kubectl/get.go [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+}
+
+func newGetCmd() *cobra.Command {
+	return getCmd
+}
+
+func init() {
+	getCmd.Flags().BoolVarP(&watch, "watch", "w", false, "watch for changes")
+	getCmd.PersistentFlags().StringVarP(&parent, "namespace", "n", "default", "parent namespace to scope request to")
+	getCmd.PersistentFlags().BoolVarP(&allNamespaces, "all-namespaces", "A", false, "list across all namespaces")
```
[the remark this answers is not in the document]
Comment (cmurphy) [reply internal/kubectl/get.go:51]: done

File: config/apiextension/apiextension.yaml [hunk 5]
```diff
@@ -0,0 +1,50 @@
+---
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  labels:
+    app: resourcelist
+  name: resourcelist
```
[the remark this answers is not in the document]
Comment (cmurphy) [reply config/apiextension/apiextension.yaml:7]: Fixed

File: config/apiextension/apiservice.yaml [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+kind: APIService
+metadata:
+  name: v1alpha2.resources.hnc.x-k8s.io
+spec:
+  group: resources.hnc.x-k8s.io
+  version: v1alpha2
+  groupPriorityMinimum: 10
+  versionPriority: 10
+  service:
+    namespace: hnc-system
```
[the remark this answers is not in the document]
Comment (cmurphy) [reply config/apiextension/apiservice.yaml:12]: This doesn't work, I'm not sure why. Maybe kustomize doesn't know about `APIService`?

File: config/rbac/resourcelist_role.yaml [hunk 7]
```diff
@@ -0,0 +1,14 @@
+---
```
[the remark this answers is not in the document]
Comment (cmurphy) [reply config/rbac/resourcelist_role.yaml:1]: Done

File: config/rbac/resourcelist_rolebinding.yaml [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+metadata:
+  name: resourcelist
+roleRef:
+  apiGroup: rbac.authorization.k8s.io
+  kind: ClusterRole
+  name: resourcelist
+subjects:
+- kind: ServiceAccount
+  name: default
+  namespace: hnc-system
```
Comment (cmurphy) [inline config/rbac/resourcelist_rolebinding.yaml:13]: FYI this allows both the new deployment and the existing HNC manager deployment to read resources, even though the latter doesn't need it. Would there be a preference to use a separate service account or separate namespace for the API extension deployment?

**Comment by rjbez17** [comment]
@adrianludwin will you be able to take another look here? If not I'll dig in this week.

File: internal/kubectl/get.go [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+	},
+}
+
+func newGetCmd() *cobra.Command {
+	return getCmd
+}
+
+func init() {
+	getCmd.Flags().BoolVarP(&watch, "watch", "w", false, "watch for changes")
+	getCmd.PersistentFlags().StringVarP(&parent, "namespace", "n", "default", "parent namespace to scope request to")
```
[the remark this answers is not in the document]
Comment (rjbez17) [reply internal/kubectl/get.go:51]: @cmurphy This seems to be the last outstanding item. The rest LGTM. Once we get this fixed up we can get it merged.

**Comment by cmurphy** [comment]
I'll rebase this and take a look at the most recent comment this week. Thanks @rjbez17 for your attention on this.

File: internal/kubectl/get.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (cmurphy) [reply internal/kubectl/get.go:51]: @rjbez17 this came from this comment https://github.com/kubernetes-sigs/hierarchical-namespaces/pull/281#issuecomment-1784121914

I laid out my thoughts on this in https://github.com/kubernetes-sigs/hierarchical-namespaces/pull/281#issuecomment-1805129703 - in short, I don't think it's feasible.

I'll resolve this comment.

**Comment by cmurphy** [comment]
Rebased

File: config/apiextension/apiextension.yaml [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+          capabilities:
+            drop: ["ALL"]
+        ports:
+        - containerPort: 7443
+          name: server
+          protocol: TCP
+      volumes:
+      - secret:
+          defaultMode: 420
+          secretName: hnc-resourcelist-apiextension
```
Comment (utam0k) [inline config/apiextension/apiextension.yaml:50]: @cmurphy Hi, it seems this secret doesn't exist in the manifest. Where is it?

**Pull request merged**